### ISSUE TITLE
Converting WAV to MP3 truncate file length

### ISSUE THREAD

**Issue by DirkFEng** [naudio/NAudio#1022]
Hi team,

I have a short example for the "problem". I create two console apps (either with .NET Framework 4.72 and the other with .NET Core 6.0).
With Nuget I add NAudio 2.1.0 and NAudio.Lama 2.1.0.
When Convert in .NET Framework the result is ok. With .NET Core the file size was (in my case) only 1/4 of the original size.
Added my example file 16 seconds long is shrinked to 4s in .NET Core version. 

[AudioFile.zip](https://github.com/naudio/NAudio/files/11259785/AudioFile.zip)

**using (var reader = new AudioFileReader("C:\\Temp\\AudioFile.wav"))
{
  using (var fileWriter = new FileStream("C:\\Temp\\Convert.mp3", FileMode.Create, FileAccess.Write))
  {
	LameConfig lc = new LameConfig();
	lc.Out
        putSampleRate = 8000;
	lc.BitRate = 16;
	using (var wtr = new LameMP3FileWriter(fileWriter, reader.WaveFormat, lc))
	{
		reader.CopyTo(wtr);
		wtr.Flush();
	}
  }
}**

Thanks in advance.
Regards
Dirk F.

**Comment by markheath** [comment]
Yes, might be best to ask someone who has LAME experience. I don't think MP3 as a format has very good support for super low sample-rates - 8kHz mono is more typically associated with telephony rather than music, which is MP3's focus. AAC or Opus would be more appropriate choices if possible.

**Comment by Corey-M** [comment]
Hi @markheath... been a minute 😄

Haven't probed too deep into this, just tried to follow along with the way `AudioFileReader` would process the file:

* Create a `WaveFileReader` on the input stream.
* If not PCM or IEEE-Float then:
  * Wrap reader in `WaveFormatConversionStream`
  * Wrap again in `BlockAlignReductionStream`

When I do that,  manually I get the same issues: `CopyTo` reads 4 seconds worth of samples and then stops. Since 4 seconds is the size of the circular buffer in `BlockAlignReductionStream` I figured removing that from the pipeline would help, and it does. Putting it back resurrects the problem. Replacing `CopyTo` with a loop that reads 2 seconds of samples at a time works fine and reads the entire file.

Not sure where the problem actually lies, whether it's in `CircularBuffer` or `BlockAlignReductionStream.Read`... been a long time since I looked at this stuff I'm well behind on your new code. (Yep, am lazy.)

**Comment by Corey-M** [comment]
OK, found the issue. [`BlockAlignReductionStream.Read()`][1] fills the circular buffer by reading the next batch of samples from the source stream and writing them to the circular buffer. The problem comes in when the requested byte count is larger than the available (or total) space in the circular buffer, in which case the extra bytes are lost. And since filling the circular buffer happens in a loop that doesn't stop until it has enough bytes to satisfy the request, it reads and discards the entire source stream before continuing.

On my test run (.NET 6.0, Windows x64) `CopyTo()` reads in 128KB chunks and repeats until the returned count is zero. The end result is that I get the first 4 seconds of the stream and the input file is left at end of input.

As I understand it the purpose of `BlockAlignReductionStream` is to eager-buffer reads from the source stream to reduce the overheads of decoding, especially when processing small numbers of samples. Decoding 4 seconds at a time is a reasonable setting for this, we just need to deal with situations where the total requested size exceeds that.

Here's one idea that seems to work:

```
protected int TrimToBlock(int value) => value - (value % sourceStream.BlockAlign);

public override int Read(byte[] buffer, int offset, int count)
{
    lock (lockObject)
    {
        int currOffset = offset;
        int currCount =  TrimToBlock(count);
        int rc;
        while (currCount > 0)
        {
            // Fill circular buffer if it has less than one block
            if (circularBuffer.Count < sourceStream.BlockAlign)
            {
                int available = TrimToBlock(circularBuffer.MaxLength - circularBuffer.Count);
                rc = sourceStream.Read(GetSourceBuffer(available), 0, available);
                if (rc > 0)
                    circularBuffer.Write(sourceBuffer, 0, rc);
            }

            // Read next block of bytes from circular buffer
            // (Do we need to trim currCount here?)
            rc = circularBuffer.Read(buffer, currOffset, currCount);
            if (rc == 0)
                break;
            currOffset += rc;
            currCount -= rc;
        }
        
        // return value is change in offset
        return currOffset - offset;
    }
}
```

This retains the principle of reading good-sized chunks from the source and filling the output array as best we can before returning. This version apparently handles `CopyTo` correctly - well, it puts the right number of bytes into a `MemoryStream`. I assume they're the right ones 😄

[1]: https://github.com/naudio/NAudio/blob/1c710c24cdc6f82947a9e7deee6faa45658404a5/NAudio.Core/Wave/WaveStreams/BlockAlignReductionStream.cs#L147C32-L148

**Comment by DirkFEng** [comment]
Hi Corey,

thx for keeping update me!
Seems there’s an solution in the near future.

Thx for your support!!

Regards
Dirk F.

Von: Corey Murtagh ***@***.***>
Gesendet: Freitag, 12. Mai 2023 03:54
An: naudio/NAudio ***@***.***>
Cc: Dirk Frischalowski ***@***.***>; Author ***@***.***>
Betreff: Re: [naudio/NAudio] Converting WAV to MP3 truncate file length (Issue #1022)


OK, found the issue. BlockAlignReductionStream.Read()<https://github.com/naudio/NAudio/blob/1c710c24cdc6f82947a9e7deee6faa45658404a5/NAudio.Core/Wave/WaveStreams/BlockAlignReductionStream.cs#L147C32-L148> fills the circular buffer by reading the next batch of samples from the source stream and writing them to the circular buffer. The problem comes in when the requested byte count is larger than the available (or total) space in the circular buffer, in which case the extra bytes are lost. And since filling the circular buffer happens in a loop that doesn't stop until it has enough bytes to satisfy the request, it reads and discards the entire source stream before continuing.

On my test run (.NET 6.0, Windows x64) CopyTo() reads in 128KB chunks and repeats until the returned count is zero. The end result is that I get the first 4 seconds of the stream and the input file is left at end of input.

As I understand it the purpose of BlockAlignReductionStream is to eager-buffer reads from the source stream to reduce the overheads of decoding, especially when processing small numbers of samples. Decoding 4 seconds at a time is a reasonable setting for this, we just need to deal with situations where the total requested size exceeds that.

Here's one idea that seems to work:

protected int TrimToBlock(int value) => value - (value % sourceStream.BlockAlign);



public override int Read(byte[] buffer, int offset, int count)

{

    lock (lockObject)

    {

        int currOffset = offset;

        int currCount =  TrimToBlock(count);

        int rc;

        while (currCount > 0)

        {

            // Fill circular buffer if it has less than one block

            if (circularBuffer.Count < sourceStream.BlockAlign)

            {

                int available = TrimToBlock(circularBuffer.MaxLength - circularBuffer.Count);

                rc = sourceStream.Read(GetSourceBuffer(available), 0, available);

                if (rc > 0)

                    circularBuffer.Write(sourceBuffer, 0, rc);

            }



            // Read next block of bytes from circular buffer

            // (Do we need to trim currCount here?)

            rc = circularBuffer.Read(buffer, currOffset, currCount);

            if (rc == 0)

                break;

            currOffset += rc;

            currCount -= rc;

        }



        // return value is change in offset

        return currOffset - offset;

    }

}

This retains the principle of reading good-sized chunks from the source and filling the output array as best we can before returning. This version apparently handles CopyTo correctly - well, it puts the right number of bytes into a MemoryStream. I assume they're the right ones 😄

**Comment by sean-m-cooper** [comment]
I'm experiencing the same issue, I believe,
In my case, I'm pulling down multiple WAV files and converting them all to MP3 before joining them together.
Here's my code.
        MemoryStream output = new();
        foreach (var audioFile in audioFiles)
        {
            Mp3FileReader reader = new(new MemoryStream(audioFile.AudioFile.AudioFileData));
            if (output.Position == 0 && reader.Id3v2Tag != null)
            {
                output.Write(reader.Id3v2Tag.RawData, 0, reader.Id3v2Tag.RawData.Length);
            }
            Mp3Frame frame;
            while ((frame = reader.ReadNextFrame()) != null)
            {
                output.Write(frame.RawData, 0, frame.RawData.Length);
            }
        }
        return await Task.FromResult(output.ToArray());
        
  Any suggestions on how to address this?
  My use case is having AI read pieces of text and then assembling them into paragraphs and larger documents. A lot of AI voice generators won't handle the size of text I need to be read in one shot.

**Comment by markheath** [comment]
Thanks for reporting, have submitted a fix for this for NAudio 3. @sean-m-cooper  I think your issue is something else - feel free to open up a separate issue if you are still having problems (although I appreciate this was quite a long time ago)
